pilot/pkg/proxy/envoy/v2/mesh_discovery.go

```diff
@@ -0,0 +1,213 @@
+//
+// Implementations of MeshDiscovery are required to be threadsafe.
+type MeshDiscovery interface {
+	// Endpoints implements EDS and returns a list of endpoints by subset for the list of supplied subsets.
+	// In Envoy's terminology a subset is service cluster.
+	Endpoints(serviceClusters []string) *xdsapi.DiscoveryResponse
+}
+
+// DiscoveryServer is Pilot's gRPC implementation for Envoy's v2 xds APIs
+type DiscoveryServer struct {
+	mu             sync.Mutex
+	mesh           MeshDiscovery
+	grpcServer     *grpc.Server
+	pendingStreams map[grpc.Stream]map[string]*chan bool
+}
+
+func init() {
+	envGRPCPort := os.Getenv("PILOT_GRPC_PORT")
+	if len(envGRPCPort) > 0 {
+		gRPCPort = envGRPCPort
+	}
+}
+
+// NewDiscoveryServer creates DiscoveryServer that sources data from Pilot's internal mesh data structures
+func NewDiscoveryServer(mesh MeshDiscovery) *DiscoveryServer {
+	// TODO for now use hard coded / default gRPC options. The constructor may evolve to use interfaces that guide specific options later.
+	// Example:
+	//		grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(uint32(someconfig.MaxConcurrentStreams)))
+	var grpcOptions []grpc.ServerOption
+
+	var interceptors []grpc.UnaryServerInterceptor
+
+	// TODO: log request interceptor if debug enabled.
+
+	// setup server prometheus monitoring (as final interceptor in chain)
+	interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	grpcOptions = append(grpcOptions, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(interceptors...)))
+
+	// get the grpc server wired up
+	grpc.EnableTracing = true
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+	out := &DiscoveryServer{mesh: mesh, grpcServer: grpcServer, pendingStreams: make(map[grpc.Stream]map[string]*chan bool)}
+
+	// TODO currently only supports EDS. CDS and other gRPC services need to be plugged-in here.
+	xdsapi.RegisterEndpointDiscoveryServiceServer(grpcServer, out)
+	return out
+}
+
+// Start runs DiscoveryServer's gRPC server on the port specified by the environment variable PILOT_GRPC_PORT.
+func (s *DiscoveryServer) Start() {
+	lis, err := net.Listen("tcp", ":"+gRPCPort)
+	if err != nil {
+		log.Errorf("mesh discovery server failed to listen on port %q: %v", gRPCPort, err)
+		return
+	}
+	// Serve is a blocking call.
+	go func(gRPCServer *grpc.Server, lis net.Listener) {
+		log.Infof("attempting to start mesh discovery server on port %q", gRPCPort)
+		if err := gRPCServer.Serve(lis); err != nil {
+			log.Errorf("mesh discovery server failed to serve on port %q: %v", gRPCPort, err)
+		}
+	}(s.grpcServer, lis)
+}
+
+/***************************  Mesh EDS Implementation **********************************/
+
+// StreamEndpoints implements xdsapi.EndpointDiscoveryServiceServer.StreamEndpoints().
+func (s *DiscoveryServer) StreamEndpoints(stream xdsapi.EndpointDiscoveryService_StreamEndpointsServer) error {
```

Comment:
So I thought about this a lot more. The current setup is fine because we'll only ever hit the code path that creates a go routine on the first request on the stream; for every other request we should hit the if statement on line 128 and skip starting a go routine.

What I'd really like to see, then, is this method restructured to make that more obvious. Something along the lines of:

```go
func StreamEndpoints(stream) error {
    cancelChannel, err := initialStreamSetup()
    defer cancelChannel.cancel()
    for {
        req, err := stream.recv()
        if err == io.EOF {
            return nil
        } else if err != nil {
            return err
        }
    }
}
```

And have `initialStreamSetup` start the go routine to write to the stream, etc. The goal is to make the common path obvious.

Alternatively, keep the existing code, but wrap the `go func` on line 140 with a `sync.Once` to make it obvious that we only ever start a goroutine a single time per node.